File: prototype/edulens-agent/src/hooks/output-guardrail.ts
// Output validation guardrail hook (PostToolUse)
// Validates agent responses after generation to catch:
// 1. Prediction language ("will definitely pass", "guaranteed")
// 2. Comparison language ("better than other students", "top 10%")
// 3. Medical/psychological advice

import type {
  HookCallback,
  HookCallbackMatcher,
  HookJSONOutput,
} from "@anthropic-ai/claude-agent-sdk";

// Prediction patterns that should be flagged
const PREDICTION_PATTERNS = [
  /will (definitely|certainly|surely) (pass|get in|be accepted|make it)/i,
  /guaranteed/i,
  /100% (chance|certain|sure)/i,
  /no doubt (she|he|they) will/i,
  /i('m| am) (certain|sure|confident) (she|he|they) will (pass|get in)/i,
  /predict (she|he|they) will/i,
  /will (pass|get in|be selected)/i,
  /chances are (very )?(high|good|excellent)/i,
  /almost certain to/i,
];

// Comparison patterns
const COMPARISON_PATTERNS = [
  /better than (other|most|many) students/i,
  /top \d+%/i,
  /above average/i,
  /below average/i,
  /ranks? (higher|lower|above|below)/i,
  /compared to (other|her|his) (peers|classmates|students)/i,
  /percentile/i,
  /outperform/i,
  /behind (other|her|his) (peers|classmates)/i,
];

// Medical/psychological advice patterns
const MEDICAL_PATTERNS = [
  /should (see|consult|visit) a (doctor|therapist|psychologist|psychiatrist)/i,
  /i (recommend|suggest) (therapy|counseling|medication)/i,
  /sounds like (adhd|autism|dyslexia|anxiety|depression)/i,
  /may have (adhd|autism|dyslexia|a learning disability)/i,
  /symptoms of/i,
  /diagnos/i,
];

export type OutputViolation = {
  type: "prediction" | "comparison" | "medical_advice";
  pattern: string;
  message: string;
};

export function checkOutputGuardrails(text: string): OutputViolation[] {
  const violations: OutputViolation[] = [];

  for (const pattern of PREDICTION_PATTERNS) {
    const match = pattern.exec(text);
    if (match) {
      violations.push({
        type: "prediction",
        pattern: match[0],
        message:
          "Response contains prediction language. " +
          "Do not make predictions about exam outcomes or admissions.",
      });
    }
  }

  for (const pattern of COMPARISON_PATTERNS) {
    const match = pattern.exec(text);
    if (match) {
      violations.push({
        type: "comparison",
        pattern: match[0],
        message:
          "Response contains comparison language. " +
          "Do not compare the student to other students or benchmarks.",
      });
    }
  }

  for (const pattern of MEDICAL_PATTERNS) {
    const match = pattern.exec(text);
    if (match) {
      violations.push({
        type: "medical_advice",
        pattern: match[0],
        message:
          "Response contains medical/psychological advice. " +
          "Do not provide medical or psychological guidance.",
      });
    }
  }

  return violations;
}

// Hook callback for the Agent SDK PostToolUse hooks system
// This is used to monitor agent text outputs after each tool use
const outputGuardrailCallback: HookCallback = async (input): Promise<HookJSONOutput> => {
  if (input.hook_event_name !== "PostToolUse") {
    return {};
  }

  const hookInput = input as typeof input & { tool_response?: unknown };
  const response = hookInput.tool_response;

  // Only check text-based tool responses
  if (typeof response !== "string") {
    return {};
  }

  const violations = checkOutputGuardrails(response);

  if (violations.length > 0) {
    const violationSummary = violations
      .map((v) => `[${v.type}] ${v.message}`)
      .join("; ");

    return {
      hookSpecificOutput: {
        hookEventName: "PostToolUse" as const,
        additionalContext:
          `OUTPUT GUARDRAIL VIOLATION DETECTED: ${violationSummary}. ` +
          "Please regenerate your response without the flagged content.",
      },
    };
  }

  return {};
};

export const outputGuardrailMatcher: HookCallbackMatcher = {
  hooks: [outputGuardrailCallback],
  timeout: 5,
};

File: edulens-agents/guardrails/output_guardrail.py
"""Output guardrails — validates agent responses to catch:
1. Prediction language ("will definitely pass", "guaranteed")
2. Comparison language ("better than other students", "top 10%")
3. Medical/psychological advice"""

import re
from dataclasses import dataclass
from typing import List

PREDICTION_PATTERNS = [
    re.compile(r"will (definitely|certainly|surely) (pass|get in|be accepted|make it)", re.IGNORECASE),
    re.compile(r"guaranteed", re.IGNORECASE),
    re.compile(r"100% (chance|certain|sure)", re.IGNORECASE),
    re.compile(r"no doubt (she|he|they) will", re.IGNORECASE),
    re.compile(r"i('m| am) (certain|sure|confident) (she|he|they) will (pass|get in)", re.IGNORECASE),
    re.compile(r"predict (she|he|they) will", re.IGNORECASE),
    re.compile(r"will (pass|get in|be selected)", re.IGNORECASE),
    re.compile(r"chances are (very )?(high|good|excellent)", re.IGNORECASE),
    re.compile(r"almost certain to", re.IGNORECASE),
]

COMPARISON_PATTERNS = [
    re.compile(r"better than (other|most|many) students", re.IGNORECASE),
    re.compile(r"top \d+%", re.IGNORECASE),
    re.compile(r"above average", re.IGNORECASE),
    re.compile(r"below average", re.IGNORECASE),
    re.compile(r"ranks? (higher|lower|above|below)", re.IGNORECASE),
    re.compile(r"compared to (other|her|his) (peers|classmates|students)", re.IGNORECASE),
    re.compile(r"percentile", re.IGNORECASE),
    re.compile(r"outperform", re.IGNORECASE),
    re.compile(r"behind (other|her|his) (peers|classmates)", re.IGNORECASE),
]

MEDICAL_PATTERNS = [
    re.compile(r"should (see|consult|visit) a (doctor|therapist|psychologist|psychiatrist)", re.IGNORECASE),
    re.compile(r"i (recommend|suggest) (therapy|counseling|medication)", re.IGNORECASE),
    re.compile(r"sounds like (adhd|autism|dyslexia|anxiety|depression)", re.IGNORECASE),
    re.compile(r"may have (adhd|autism|dyslexia|a learning disability)", re.IGNORECASE),
    re.compile(r"symptoms of", re.IGNORECASE),
    re.compile(r"diagnos", re.IGNORECASE),
]


@dataclass
class OutputViolation:
    type: str  # "prediction", "comparison", "medical_advice"
    pattern: str
    message: str


def check_output_guardrails(text: str) -> List[OutputViolation]:
    """Check agent output text for guardrail violations."""
    violations = []

    for pattern in PREDICTION_PATTERNS:
        match = pattern.search(text)
        if match:
            violations.append(OutputViolation(
                type="prediction",
                pattern=match.group(0),
                message="Response contains prediction language. Do not make predictions about exam outcomes or admissions.",
            ))

    for pattern in COMPARISON_PATTERNS:
        match = pattern.search(text)
        if match:
            violations.append(OutputViolation(
                type="comparison",
                pattern=match.group(0),
                message="Response contains comparison language. Do not compare the student to other students or benchmarks.",
            ))

    for pattern in MEDICAL_PATTERNS:
        match = pattern.search(text)
        if match:
            violations.append(OutputViolation(
                type="medical_advice",
                pattern=match.group(0),
                message="Response contains medical/psychological advice. Do not provide medical or psychological guidance.",
            ))

    return violations
